src/Data/TypeRepMap/Internal.hs

```diff
@@ -266,25 +270,32 @@ hoistWithKey f (TypeRepMap as bs ans ks) = TypeRepMap as bs newAns ks
     withTr t = withTypeable t f
 {-# INLINE hoistWithKey #-}
 
--- | The union of two 'TypeRepMap's using a combining function.
+-- | The union of two 'TypeRepMap's using a combining function for conflicting entries
 unionWith :: forall f. (forall x. Typeable x => f x -> f x -> f x) -> TypeRepMap f -> TypeRepMap f -> TypeRepMap f
-unionWith f m1 m2 = fromTriples
-                  $ toTripleList
-                  $ Map.unionWith combine
-                                  (fromTripleList $ toTriples m1)
-                                  (fromTripleList $ toTriples m2)
+unionWith f ma mb = do
```

Comment:
You wrote an excellent description of how `unionWith` works! Could you, please put into the Haddock documentation for this function as well? E.g. under the _Implementation description_ note or something like that. And the performance analysis is also quite useful :slightly_smiling_face: